src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -73,20 +79,72 @@ public ImmutableArray<DiagnosticData> GetBuildErrors()
         public bool IsSupportedDiagnosticId(ProjectId projectId, string id)
             => BuildInprogressState?.IsSupportedDiagnosticId(projectId, id) ?? false;
 
+        private void OnBuildProgressChanged(InProgressState? state, BuildProgress buildProgress)
+        {
+            if (state != null)
+            {
+                _lastBuiltResult = state.GetBuildErrors();
+            }
+
+            RaiseBuildProgressChanged(buildProgress);
+        }
+
         public void ClearErrors(ProjectId projectId)
         {
             // capture state if it exists
             var state = BuildInprogressState;
 
-            _taskQueue.ScheduleTask(nameof(ClearErrors), () =>
+            _taskQueue.ScheduleTask(nameof(ClearErrors), async () =>
             {
-                // this will get called if the project is actually built by "build" command.
-                // we track what project has been built, so that later we can clear any stale live errors
-                // if build give us no errors for this project
-                state?.Built(projectId);
+                if (state == null)
+                {
+                    await ClearErrorsCoreAsync(projectId, _workspace.CurrentSolution, state).ConfigureAwait(false);
+                }
+                else
+                {
+                    if (state.WereProjectErrorsCleared(projectId))
+                    {
+                        return;
+                    }
+
+                    var solution = state.Solution;
 
-                ClearProjectErrors(state?.Solution ?? _workspace.CurrentSolution, projectId);
+                    // Clear errors for the given project.
+                    await ClearErrorsCoreAsync(projectId, solution, state).ConfigureAwait(false);
+
+                    // Additionally, clear errors for all projects that transitively depend on this project.
+                    // Otherwise, fixing errors in core projects in dependency chain will leave back stale diagnostics in dependent projects.
+                    var transitiveProjectIds = solution.GetProjectDependencyGraph().GetProjectsThatTransitivelyDependOnThisProject(projectId);
+                    foreach (var projectId in transitiveProjectIds)
+                    {
+                        if (state.WereProjectErrorsCleared(projectId))
+                        {
+                            continue;
+                        }
+
+                        await ClearErrorsCoreAsync(projectId, solution, state).ConfigureAwait(false);
+                    }
+                }
             }, CancellationToken.None);
+
+            return;
+
+            async Task ClearErrorsCoreAsync(ProjectId projectId, Solution solution, InProgressState? state)
+            {
+                Debug.Assert(state == null || !state.WereProjectErrorsCleared(projectId));
+
+                // Clear build-only errors for project
+                ClearBuildOnlyProjectErrors(solution, projectId);
```

Comment:
comment is not helpful.  it just restates the code.  would love more of an explanation/intuition style of comments here.  applies to the rest of hte comments.